### PR TITLE
WIP: Faster implementation of linreg

### PR DESCRIPTION
**Pull request by gabrielgellner**
Move from using the general matrix division operator to an explicit loop to calculate the OLS without extra memory allocation.

### CODE REVIEW

**Comment by gabrielgellner** [comment]
I have changed the implementation of linreg following the discussion with @andreasnoack on julia-users. I would appreciate any feedback on the implementation, including what is the general preference for the return type, currently I am returning an Array but @andreasnoack  has indicated that a Tuple might be preferred for potential speed increase (which I currently can't find evidence of) as well as being a more natural output, as the user might prefer to only use one of the two returned coeffcients. I like being able to do vector arithmetic on the output, but the longer I think about it, the more agnostic I am becoming.

I have also tried to ensure that this pull request is rebased and all the other good git etiquette, but I am not sure I have done it correctly. Tips, feedback appreciated.

**Comment by andreasnoack** [comment]
Thanks for the non-`\` version. I think it would be best to use `cov` even though it's not as fast as it should be. The reason is partly that we shouldn't reimplement an existing function. Instead, we should make the slow `cov` faster. Maybe, we should open an issue about the speed of `cov`. When it gets fixed then `linreg` will automatically benefit.

The other reason is that `cov` and `var` should (hopefully) handle all kinds of input well. E.g. your implementation will promote the result to `Float64` even for `Float32` input and it can be a bit painful to get the accumulator in the loop intialized in a generic way that allows for general input.

**Comment by gabrielgellner** [comment]
I can easily move back to a `cov/var` version, though I wonder if it is worth it, a lot of the speedup feels like it stems from taking advantage of combining operations that are shared between `var` and `cov` during a single traversal of the data. Do you think this is just a result of `cov` being slow? 

Wouldn't having to use two functions for this lead to having the extra traversal of `x` and `y` (once for calculating cov and once for calculating var) no matter how fast the internal algorithm is? Is there a way to combine the inner iteration between the two functions they way my algorithm is across function boundaries?

Your point about the type promotion is very interesting. I hadn't thought of that. I am struck by the difficulty of doing this correctly, especially if, like both my proposals allow for using `UnitRange`/`LinRange` objects. I am not sure what to do about this. Feels like a tonne of special casing would be needed!

Finally do you have any final insight on why the `\\` version versus the `cov/var` version should be fundamentally more efficient? Maybe this is all, like you mention, over optimizing a specific case when instead `\\` should be made faster instead?

File: .gitignore [hunk 1]
```diff
@@ -24,3 +24,4 @@
 *.ji
 
 .DS_Store
+get_toolchain.log
```
Comment (StefanKarpinski) [inline .gitignore]: I don't think this should be in here.

File: base/linalg/generic.jl [hunk 2]
```diff
@@ -417,7 +417,49 @@ istriu(x::Number) = true
...
+    # b1 = cov(X, Y)/var(X)
+    # b0 = mean(Y) - b1*mean(X)
+    # Calculate the cov and var using the fewest iterations of the inputs possible
+    n = length(x)
+    @assert n == length(y)
+    mx = mean(x) # need to know this at each step do the implicit loop over x first
+    tot_dx = 0.0
+    tot_y = 0.0
+    tot_xy = 0.0
+    for i in 1:n
```
Comment (KristofferC) [inline base/linalg/generic.jl]: can use `@inbounds` since you checked the length

File: .gitignore (hunk 1 above)
Comment (gabrielgellner) [inline .gitignore]: I agree. I can not for the life of me figure out how to remove that file from the pull request (or requests as it seems to have combined both). I really just want the changes to `linalg/generics.jl` and `tests/linalg/generics.jl`. But man every time I try to follow the guides on rebasing or amending ... it nukes everything. Hardest 10 line change to code I have had to do. I am might be to dumb to use git :(

File: base/linalg/generic.jl (hunk 2 above)
Comment (gabrielgellner) [inline base/linalg/generic.jl]: Do I just put the `@inbounds` in front of the for loop? Doesn't seem to change any of my timings, but seems harmless.

**Comment by gabrielgellner** [comment]
Okay. I have cleaned up all the nonsense from the previous pulls. No more merge garbage, or messing with the upstream .gitignore.

I will wait a bit to see if there are anymore opinions on the two versions of the code. And if not I will make the changes suggested by @andreasnoack as he has a better feeling for the statistical functions in Base.

File: .gitignore (hunk 1 above)
Comment (tkelman) [inline .gitignore]: I'm not opposed to adding this file to gitignore, it comes from some windows-specific setup scripts. It could maybe be a separate commit or PR but rebasing it out can be slightly complicated.

Comment (gabrielgellner) [inline .gitignore]: So so complicated. I was about to give up programming and go back to a slide rule...

File: base/linalg/generic.jl [hunk 3]
```diff
@@ -417,48 +417,38 @@ istriu(x::Number) = true
...
 
-- `x::AbstractVector{T <: Number}` - vector of independent variables
-- `y::AbstractVector{S <: Number}` - vector of dependent variables
+- `x::AbstractVector` - vector of independent variables
+- `y::AbstractVector` - vector of dependent variables
 
 Output:
 
-- `[b0, b1]` - the coefficients for the line of best fit such that yhat = b0 + b1*x
+- `(a, b)` - the coefficients for the line of best fit satisfying a + b*x
```
Comment (ararslan) [inline base/linalg/generic.jl]: I actually prefer the previous wording and use of `b0` and `b1`. If it were me, I would just change the square brackets to parentheses and leave the description unchanged.

File: base/linalg/generic.jl [hunk 4]
```diff
@@ -417,48 +417,38 @@ istriu(x::Number) = true
...
 See also:
 
 `\\`, `cov`, `std`, `mean`
 
 """
-function linreg{T <: Number, S <: Number}(x::AbstractVector{T}, y::AbstractVector{S})
+function linreg(x::AbstractVector, y::AbstractVector)
     # Least squares given
-    # Y = b0 + b1*X
+    # Y = a + b*X
```
Comment (ararslan) [inline base/linalg/generic.jl]: Out of curiosity, why change to `a` and `b`? Personally I've always found the subscripted beta notation for regression coefficients much clearer than `a` and `b`.

File: base/linalg/generic.jl [hunk 5]
```diff
@@ -417,48 +417,38 @@ istriu(x::Number) = true
...
-        tot_dx += (x[i] - mx)^2
-        tot_y += y[i]
-        tot_xy += x[i]*y[i]
-    end
-    b1 = (tot_xy - mx*tot_y)/tot_dx # a 1\n cancels in the top and bottom
-    b0 = tot_y/n - b1*mx
-    return [b0, b1]
+    # b = cov(X, Y)/var(X)
+    # a = mean(Y) - b*mean(X)
+    size(x) == size(y) || throw(DimensionMismatch())
```
Comment (ararslan) [inline base/linalg/generic.jl]: It might be nice to provide at least a minimal decription in the `DimensionMismatch` exception constructor.

File: test/linalg/generic.jl [hunk 6]
```diff
@@ -65,26 +65,34 @@ let X = [3  9   5;
...
+y = [3; 7; 10]
+@test_throws MethodError linreg(x, y) 
+
 # check (UnitRange, Array)
 x = 1:12
 y = [5.5; 6.3; 7.6; 8.8; 10.9; 11.79; 13.48; 15.02; 17.77; 20.81; 22.0; 22.99]
 @test_approx_eq linreg(x,y) [2.5559090909090867, 1.6960139860139862]
 @test_approx_eq linreg(sub(x,1:6),sub(y,1:6)) [3.8366666666666642,1.3271428571428574]
 
-##The following tests will not require collect once issue #16520 is fixed
```
Comment (ararslan) [inline test/linalg/generic.jl]: Has that been fixed? The issue is still open.

File: base/linalg/generic.jl (hunk 3 above)
Comment (gabrielgellner) [inline base/linalg/generic.jl]: I am happy to use either, see if there is any consensus. I changed this so that it was consistent with the example given in the `Base.rst` file. Changing to the parentheses is meant to be consistent with the code soon, I just can't figure out how to use the `@test_approx_equal_eps` with a tuple return instead of the array...

File: base/linalg/generic.jl (hunk 4 above)
Comment (gabrielgellner) [inline base/linalg/generic.jl]: Yeah my brain clearly gravitates to the beta notation, but when looking over the existing code I realized I had introduced inconsistency with other parts of the code so I changed this.

File: base/linalg/generic.jl (hunk 5 above)
Comment (gabrielgellner) [inline base/linalg/generic.jl]: I will look into this. Was just following the code from the recent changes to `quantile`

File: test/linalg/generic.jl (hunk 6 above)
Comment (gabrielgellner) [inline test/linalg/generic.jl]: The issue is still a problem, but since this code no longer uses the `\` operator it is not effected and the tests work as intended, hence the removal of the `collect` hacks.

File: test/linalg/generic.jl [hunk 7]
```diff
@@ -65,11 +65,52 @@ let X = [3  9   5;
...
+# check (UnitRange, Array)
+x = 1:12
 y = [5.5; 6.3; 7.6; 8.8; 10.9; 11.79; 13.48; 15.02; 17.77; 20.81; 22.0; 22.99]
 @test_approx_eq linreg(x,y) [2.5559090909090867, 1.6960139860139862]
 @test_approx_eq linreg(sub(x,1:6),sub(y,1:6)) [3.8366666666666642,1.3271428571428574]
 
+# check (LinSpace, UnitRange)
+x = linspace(1.0, 12.0, 100)
+y = -100:-1
+@test_approx_eq linreg(x, y) [-109.0, 9.0]
```
Comment (andreasnoack) [inline test/linalg/generic.jl]: In case you return a tuple, you can do `[linreg(x,y)...]` to convert it to an array here.

Comment (gabrielgellner) [inline test/linalg/generic.jl]: Great thanks. I have changed the return type to a tuple as you have suggested, and the tests to reflect this.

**Comment by andreasnoack** [comment]
LGTM. Do you consider it ready? If so then please [squash the commits](https://github.com/ginatrapani/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit) (migth be little fight with git the first time you do it but you'll have to go through this process sooner or later).

**Comment by KristofferC** [comment]
Why not squash with github ui?

**Comment by tkelman** [comment]
Is this docstring intended to replace the old one? Don't think that's ready yet, should run `make docs` to check that it's propagating into the rst.

**Comment by andreasnoack** [comment]
@KristofferC I didn't know about that option. So if an owner merges the PR, he can do the squash?

**Comment by KristofferC** [comment]
Yes, when you press merge you can chose to squash it (and edit the commit message if you chose). It will also avoid the merge commit.

**Comment by tkelman** [comment]
See https://github.com/blog/2141-squash-your-commits - when you click "Merge pull request," but before you click "Confirm merge," there's an option you can select. It's sticky though, so keep an eye on it since you might not want to squash commits from every single PR, for those that are broken into good separate (all passing) pieces that may be worth bisecting within later.

**Comment by gabrielgellner** [comment]
@tkelman There was no docstring before, but there is an entry in `base/docs/helpdb/Base.jl` which I have not changed. I can run the `make docs`, will this create a new file I need to commit?

@andreasnoack I am happy to try and figure out how to do the squash if that is preferred ... learned more about git then I ever wanted in the last couple of days ...

**Comment by tkelman** [comment]
The contents of `base/docs/helpdb/Base.jl` are docstrings, they're just all in one leftover file from the doc migration last year. They're supposed to gradually get moved inline to docstrings at the function signature like the new one you're adding. Having a docstring in `helpdb/Base.jl` and inline for the same function is bad, I'm not sure which will show up when you query the repl help - they might both get concatenated which would be redundant here.

So I'd recommend deleting the one from `helpdb/Base.jl`, maybe incorporating anything that's worth keeping into your new docstring. Then you run `make docs` and it should translate your markdown inline docstring into rst and populate it in the "generated from Julia source" section of the rst manual. Commit that change and add it here, so the next person who runs `make docs` doesn't get confused by your change which is unrelated to whatever they're working on. It's a messy system but we're aiming to eventually move away from rst and sphinx altogether. Next release cycle, hopefully.

**Comment by gabrielgellner** [comment]
@tkelman okay I have merged the docstring and the information in `helpdb`. Before I push this to get feedback are there any guidelines for docstrings? I have followed the example given in `linalg/generics.jl:normalize` which seem to follow the scipy format, but when I look elsewhere this does not seem to be followed. Is there a preferred layout or is it function to function?

@andreasnoack fixing up the docstrings I see the case for weighted `linreg`, do we want to change the algorithm on this or do you think it is fine as is?

**Comment by tkelman** [comment]
ref https://github.com/JuliaLang/julia/pull/15136 for the most recent discussion on docstring guidelines

**Comment by andreasnoack** [comment]
I'd forgotten the weighted version. Actually, it shouldn't be here. We have moved all the weighted statistics functions to `StatsBase.jl` so that is where it belongs. In `StatsBase.jl`, we could also simply use the weighted variance and covariance functions there. Would you mind opening a PR against `StatsBase.jl` with a similar version

**Comment by gabrielgellner** [comment]
@tkelman I am not sure what is happening but the `rst` file is not being updated when I do `make docs` I have tried doing a `make clean` first but the section in `doc/stdlib/linalg.rst` is not reflecting the changes I have made to the docstrings and is still reflecting the deleted section in `helpdb` as well as having a information from the deleted weighted `linreg` method. Not sure what I am doing wrong. The help from the REPL is all good now.

@andreasnoack I have removed the weighted version, once I get the documentation stuff fixed up I will open a pull request in `StatsBase.jl`. Thanks for all the feedback.

**Comment by gabrielgellner** [comment]
Okay I believe everything is good to go! Let me know if all the recent changes look good and I can figure out how to squash the commits.

**Comment by ararslan** [comment]
`git reset --soft HEAD~X` where `X` is the number of commits you want to squash (i.e. squash the `X` most recent commits into one), then `git commit` and write a new commit message. :smile:

**Comment by KristofferC** [comment]
I always do `git rebase -i HEAD~X` and shose squash

**Comment by gabrielgellner** [comment]
So I have my local copy of the branch rebased and squashed. But I don't understand how to push it over this current pull request.

I tried
git push -f origin

but I get strange error messages about no the name of the upstream branch. Am I missing something simple?

**Comment by tkelman** [comment]
You need to force push over the `linreg-updates` branch of the remote that corresponds to your fork. Provide both the remote name and the branch name to push.

**Comment by gabrielgellner** [comment]
OMG I hate git. Disregard all the commit noise ...

**Comment by gabrielgellner** [comment]
Okay ready to rock.

File: base/linalg/generic.jl [hunk 8]
```diff
@@ -417,12 +417,40 @@ istriu(x::Number) = true
 istril(x::Number) = true
 isdiag(x::Number) = true
 
-linreg{T<:Number}(X::AbstractVector{T}, y::AbstractVector{T}) = [ones(T, size(X,1)) X] \ y
+"""
+    linreg(x, y)
+
+Perform simple linear regression using Ordinary Least Squares. Return `a` and `b` such
```
Comment (tkelman) [inline base/linalg/generic.jl]: Returns

**Comment by andreasnoack** [comment]
Thanks for the contribution.
